packages/cli/src/modules/insights/insights-pruning.service.ts

```diff
@@ -0,0 +1,90 @@
+import { Service } from '@n8n/di';
+import { Logger } from 'n8n-core';
+
+import { Time } from '@/constants';
+
+import { InsightsByPeriodRepository } from './database/repositories/insights-by-period.repository';
+import { InsightsConfig } from './insights.config';
+
+@Service()
+export class InsightsPruningService {
+	private pruneInsightsTimer: NodeJS.Timeout | undefined;
+
+	private isStopped = false;
+
+	private readonly delayBetweenRetries = Time.seconds.toMilliseconds * 10;
+
+	private readonly maxRetries = 3;
+
+	constructor(
+		private readonly insightsByPeriodRepository: InsightsByPeriodRepository,
+		private readonly config: InsightsConfig,
+		private readonly logger: Logger,
+	) {
+		this.logger = this.logger.scoped('insights');
+	}
+
+	get isPruningEnabled() {
+		return this.config.maxAgeDays > -1;
+	}
+
+	startPruningTimer() {
+		if (!this.isPruningEnabled) {
+			return;
+		}
+
+		this.clearPruningTimer();
+		this.isStopped = false;
+		this.scheduleNextPrune();
+		this.logger.debug(`Insights pruning every ${this.config.pruneCheckIntervalHours} hours`);
+	}
+
+	private clearPruningTimer() {
+		if (this.pruneInsightsTimer !== undefined) {
+			clearTimeout(this.pruneInsightsTimer);
+			this.pruneInsightsTimer = undefined;
+		}
+	}
+
+	stopPruningTimer() {
+		this.isStopped = true;
+		this.clearPruningTimer();
+		this.logger.debug('Stopped Insights pruning');
+	}
+
+	private scheduleNextPrune() {
+		if (this.isStopped) return;
+
+		this.pruneInsightsTimer = setTimeout(async () => {
+			await this.pruneInsights();
+			this.scheduleNextPrune(); // reschedule after execution
+		}, this.config.pruneCheckIntervalHours * Time.hours.toMilliseconds);
+	}
+
+	async pruneInsights() {
+		for (let attempt = 1; attempt <= this.maxRetries; attempt++) {
```

Comment:
Insights pruning tries strictly 3 times with a 10s delay, while executions pruning reschedules the next pruning cycle almost immediately, within 1s. Is there a reason to handle insights pruning failures different from executions pruning failures? Also are we okay with giving up after 3 tries and causing the next cycle after 24h to be an even bigger deletion?

Reply:
Yeah I guess you're right, no reason to stop trying. I implemented the same mechanism as for executions pruning